### PR TITLE
ECG Rpeak DataStream created

### PR DESCRIPTION
**Pull request by aungkonazim**
Respected Dr. Hnat and Dr. Ali,

Please check the main.py ,cStress.py which are changed a bit to run the program
The functions are in 'memphisdataprocessor/preprocessor/' directory
Please check the ecg_processor.py and rpeak_detect.py files

Please let me know if I am lacking in any technical abilities to upload the functions through to you.
Thank you

### CODE REVIEW

**Review by twhnat — CHANGES_REQUESTED** [review]
Please take a look the codacy-bot's comments and resolve these issues.  It enforces good coding practices that we would incorporated into the codebase.

File: cerebralcortex/kernel/datatypes/datastream.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (twhnat) [inline cerebralcortex/kernel/datatypes/datastream.py]: Extraneous, please start files with the license block

File: cerebralcortex/kernel/datatypes/datastream.py [hunk 2]
```diff
@@ -28,16 +29,17 @@
...
 class DataStream:
     def __init__(self,
-                 user: UUID,
-                 id: int = None,
-                 data: RDD = None,
-                 processing: dict = None,
-                 sharing: dict = None,
-                 metadata: Metadata = Metadata(),
-                 windows: list = None):
+                    user: UUID,
```
Comment (twhnat) [inline cerebralcortex/kernel/datatypes/datastream.py]: Please revert these changes, parameters are expected to be aligned vertically as with the original file.

File: cerebralcortex/kernel/datatypes/datastream.py [hunk 3]
```diff
@@ -59,10 +62,13 @@ def __init__(self,
...
         return self._datapoints
 
+    def set_datapoints(self, rdd):
+        pass
+
 
     def add_window(self,
                    window_rdd: RDD,
-                   window_metadata: dict):
+                               window_metadata: dict):
```
Comment (twhnat) [inline cerebralcortex/kernel/datatypes/datastream.py]: Same as above

File: cerebralcortex/kernel/datatypes/datastream.py [hunk 4]
```diff
@@ -86,10 +91,12 @@ def update_metadata(self, metadata):
...
 
     @classmethod
     def from_datastream(self, datastream):
-        result = DataStream(user=datastream._user,
-                            data=datastream._datapoints)
-
-        return result
+        result = DataStream(cerebralcortex=datastream.get_cerebralcortex(),
+                            user=datastream._user)
+        return result
```
Comment (twhnat) [inline cerebralcortex/kernel/datatypes/datastream.py]: Newline expected at the end of the file

File: main.py [hunk 5]
```diff
@@ -32,43 +32,35 @@
 
 CC = cerebralcortex.CerebralCortex(master="local[*]", name="Memphis cStress Development App")
 
-basedir = "/Users/hnat/Desktop/data/"
-
+basedir = "/home/aungkon/Desktop/data/"
```
Comment (twhnat) [inline main.py]: Your code changes should be made within the cStress function, this breaks other logic by commenting out code.  If necessary, please create your own test file on which to run

File: memphisdataprocessor/cStress.py [hunk 6]
```diff
@@ -26,7 +26,8 @@
 from cerebralcortex.kernel.datatypes.window import Window
 from cerebralcortex.kernel.window import window
 from memphisdataprocessor.alignment import timestamp_correct, timestampCorrectAndSequenceAlign
-
+from memphisdataprocessor.preprocessor.ecg_processor import compute_rr_datastream
+import numpy as np
```
Comment (twhnat) [inline memphisdataprocessor/cStress.py]: This is not part of the requirements.txt file along with scipy.  This has caused the CI server to fail

File: memphisdataprocessor/cStress.py [hunk 7]
```diff
@@ -102,11 +103,12 @@ def joinDataStreams(datastreams, metadata):
     pass
 
 
-def cStress(rawecg: DataStream,
-            rawrip: DataStream,
-            rawaccelx: DataStream,
-            rawaccely: DataStream,
-            rawaccelz: DataStream) -> DataStream:
+def cStress(x):
```
Comment (twhnat) [inline memphisdataprocessor/cStress.py]: You have modified the cStress method, this is not to be done as it needs to process all five datstreams.  Please correct this back to the original call parameters

File: memphisdataprocessor/preprocessor/ecg_processor.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+# Copyright (c) 2016, MD2K Center of Excellence
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/ecg_processor.py]: Copyright year should be 2017

File: memphisdataprocessor/preprocessor/ecg_processor.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+from cerebralcortex.kernel.datatypes.datastream import DataStream
+from memphisdataprocessor.preprocessor.rpeak_detect import detect_rpeak
+import numpy as np
+import uuid
+
+def classify_ecg_window(window):
+    '''
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/ecg_processor.py]: Please use the Intellij documentation template to properly comment methods and parameters.  See the other files as reference

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
+# Copyright (c) 2016, MD2K Center of Excellence
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: Copyright year needs updated

File: main.py [hunk 11]
```diff
@@ -32,43 +32,35 @@
...
+datastream_array=[]
 for i in range(1, 2):
     participant = "SI%02d" % i
 
     participant_UUID = uuid.uuid4()
     try:
-        ecgRDD = CC.readfile(find(basedir, {"participant": participant, "datasource": "ecg"})).map(
```
Comment (twhnat) [inline main.py:51]: These should not be commented out

File: memphisdataprocessor/ecgFeaturepre.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+
+def window_temp(datastream:DataStream, windowsize:int):
+    rdd = datastream.get_datapoints()
+    result = rdd.map(lambda x: (int(x.get_timestamp_epoch() / 1000), (x.get_timestamp_epoch()/1000, x.get_sample())))
+    result = result.map(lambda x:(x[0]-x[0]% windowsize,[x[1]])).reduceByKey(lambda a,b:a+b)
+    return result
+
+
+def classifyECGwindow(datapoints):
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Needs documentation and parameterized

File: memphisdataprocessor/ecgFeaturepre.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    if np.median(np.abs(np.diff(values))) > 50:
+        return False
+    return True
+
+def filterBadECG(datastream: DataStream, windowsize: int = 2000):
+    KVrdd = window_temp(datastream, windowsize)
+    labels=KVrdd.map(lambda x:(x[0],x[1],classifyECGwindow(x[1]))).filter(lambda x:x[2]== True)  # [(TS, False), (TS2, True) ... ]
+    labels=labels.map(lambda x:(1,x[1])).reduceByKey(lambda a,b:a+b).map(lambda x:x[1]).flatMap(lambda x:x)
+    return labels
+    # meta = {x[1]}
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Remove unused commented code

File: memphisdataprocessor/ecgFeaturepre.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    values = np.array([i[1] for i in datapoints])
+    if max(values) - min(values) < 200:
+        return False
+    if max(values) > 4000:
+        return False
+    if np.median(np.abs(np.diff(values))) > 50:
+        return False
+    return True
+
+def filterBadECG(datastream: DataStream, windowsize: int = 2000):
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Document this function

File: memphisdataprocessor/ecgFeaturepre.py [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    labels=KVrdd.map(lambda x:(x[0],x[1],classifyECGwindow(x[1]))).filter(lambda x:x[2]== True)  # [(TS, False), (TS2, True) ... ]
+    labels=labels.map(lambda x:(1,x[1])).reduceByKey(lambda a,b:a+b).map(lambda x:x[1]).flatMap(lambda x:x)
+    return labels
+    # meta = {x[1]}
+    # labeledWindows = labels.map(lambda x: Window(x[0], x[0] + windowsize, meta))
+    # compressedWindows = mergeWindows(labeledWindows)
+    # return compressedWindows
+
+
+def rr_ave_update(Rpeak_temp1, rr_ave2):
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Document this function

File: memphisdataprocessor/ecgFeaturepre.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    R=[0]+Rpeak_temp1
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: What is the intended purpose of this statement?

Comment (twhnat) [reply memphisdataprocessor/ecgFeaturepre.py]: What does R mean?

File: memphisdataprocessor/ecgFeaturepre.py [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    peak_interval = np.diff(R)
+    if len(peak_interval) <= 7:
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Hard coded parameters should not be part of these functions.  Constants need to be moved to parameters

File: memphisdataprocessor/ecgFeaturepre.py [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+        return rr_ave2
+    else:
+        x=sum(peak_interval[-8:]) / 8
+        return x
+
+def detect_Rpeak_firststep(ecgfilteredrdd,fs:int=64,threshold:float=0.5):
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Document this function

File: memphisdataprocessor/ecgFeaturepre.py [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+    print(ecgfilteredrdd.count())
+    x=ecgfilteredrdd.collect()
+    sample=[i[1] for i in x]
+    timestamp=[i[0] for i in x]
+    #this program implements the Pan Tomkins algorithm on ECG signal to detect the R peaks
+
+    f = 2 / fs
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Variables need more descriptive names

File: memphisdataprocessor/ecgFeaturepre.py [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+    blackmanWinlen = np.ceil(fs / 5)
+    h_I = np.blackman(blackmanWinlen) # blackman window
+    y5 = signal.convolve(y4, h_I, 'same') #moving window Integration of squared derivative signal
+    y5 = y5 / np.percentile(y5, 90)
+
+    thr1 = threshold
+
+
+def ecgpre(rawecg: DataStream):
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Document this function

File: memphisdataprocessor/ecgFeaturepre.py [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    from test import CC
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: Imports do not belong inside of functions.

File: memphisdataprocessor/ecgFeaturepre.py [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+    global CC
```
Comment (twhnat) [inline memphisdataprocessor/ecgFeaturepre.py]: No global variables

File: memphisdataprocessor/preprocessor/ecg_processor.py [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    values = np.array([i[1] for i in window])
+    if max(values) - min(values) < 200:
+        return False
+    if max(values) > 4000:
+        return False
+    if np.median(np.abs(np.diff(values))) > 50:
+        return False
+    return True
+
+def filter_bad_ecg(ecg_array,fs):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/ecg_processor.py]: Document this function

File: memphisdataprocessor/preprocessor/ecg_processor.py [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+    for i in range(0,len(ecg_array)-window_length-1,window_length):
+        index = [k for k in range(i,i+window_length)]
+        if classify_ecg_window(ecg_array[index])==True:
+            for x in ecg_array[index]:
+                ecg_filtered_array.append((x[0],x[1]))
+
+    return ecg_filtered_array
+
+
+def compute_rr_datastream(ecg:DataStream,fs:float)->DataStream:
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/ecg_processor.py]: Document this function

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import numpy as np
+from scipy import signal
+
+def rr_interval_update(rpeak_temp1, rr_ave):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: Document this function

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+    R=[0]+rpeak_temp1
+    peak_interval = np.diff(R)
+    if len(peak_interval) <= 7:
+        return rr_ave
+    else:
+        x=np.sum(peak_interval[-8:]) / 8
+        return x
+
+
+def compute_moving_window_int(sample,fs,blackmanWinlen):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: Document this function

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+    '''
+
+    :param sample: ecg sample array
+    :param fs: sampling frequency
+    :param blackmanWinlen: length of the blackman window on which to compute the moving window integration
+
+    :return: the Moving window integration of the sample array
+    '''
+    f = 2 / fs
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: Variables need more descriptive names

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+    mov_win_int_signal = mov_win_int_signal / np.percentile(mov_win_int_signal, 90)
+    return mov_win_int_signal
+
+def check_peak(y,i):
+    '''
+    This is a function to check the condition of a simple peak of signal y in index i
+    '''
+    return y[i - 2] < y[i - 1] and y[i - 1] < y[i] and y[i] >= y[i + 1] and y[i + 1] > y[i + 2]
+
+def compute_r_peaks(rpeak_temp1,pkt_loc,sig_lev,noise_lev,thr1,thr2,rr_ave,mov_win_int_signal,valuepks,pkt):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: Document this function

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 29]
```diff
@@ -0,0 +1,195 @@
...
+    y4 = y3 ** 2 #squared derivative
+    y4 = y4 / np.percentile(y4, 90)
+
+    blackmanWinlen = np.ceil(fs / 5)
+    h_I = np.blackman(blackmanWinlen) # blackman window
+    mov_win_int_signal = signal.convolve(y4, h_I, 'same') #moving window Integration of squared derivative signal
+    mov_win_int_signal = mov_win_int_signal / np.percentile(mov_win_int_signal, 90)
+    return mov_win_int_signal
+
+def check_peak(y,i):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py:83]: Document this function

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 30]
```diff
@@ -0,0 +1,195 @@
...
+                i = pkt_loc[-1] + 1
+            else:
+                thr1 = noise_lev + 0.25 * (sig_lev - noise_lev)
+                thr2 = 0.5 * thr1
+                i = i + 1
+        else:
+            if mov_win_int_signal[pkt[i]] >= thr1 and mov_win_int_signal[pkt[i]] < 3 * sig_lev:
+                rpeak_temp1.append(pkt[i])
+                pkt_loc.append(i)
+                sig_lev = 0.125 * mov_win_int_signal[pkt[i]] + 0.875 * sig_lev
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: An EMWA function should be created to abstract this out

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 31]
```diff
@@ -0,0 +1,195 @@
...
+            elif mov_win_int_signal[pkt[i]] < thr1 and mov_win_int_signal[pkt[i]] > thr2:
+                noise_lev = 0.125 * mov_win_int_signal[pkt[i]] + 0.875 * noise_lev
+            thr1 = noise_lev + 0.25 * (sig_lev - noise_lev)
+            thr2 = 0.5 * thr1
+            i = i + 1
+    return rpeak_temp1
+
+def remove_close_peaks(rpeak_temp1,sample,fs):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: Document this function

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 32]
```diff
@@ -0,0 +1,195 @@
...
+    '''
+    This function removes one of two peaks from the detected R peaks
+    if difference among them is less than the minimum possible for two consecutive R peaks
+    '''
+    def check_min(sample,rpeak_temp1,i):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py:169]: Document this function

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 33]
```diff
@@ -0,0 +1,195 @@
...
+    while len(temp_ind) != 0:
+        temp = np.diff(rpeak_temp1)
+        temp_ind=[check_min(sample,rpeak_temp1,i) for i in range(len(temp)) if temp[i]<.5*fs]
+        count=0
+        for i in temp_ind:
+            rpeak_temp1.remove(rpeak_temp1[i-count])
+            count=count+1
+    return rpeak_temp1
+
+def confirm_peaks(rpeak_temp1,sample,fs):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: Document this function

File: memphisdataprocessor/preprocessor/rpeak_detect.py [hunk 34]
```diff
@@ -0,0 +1,195 @@
...
+    finds the maximum in a range of fs/10 of the detected peak location and assigns it to be the peak    '''
+    count=1
+    for i in range(1,len(rpeak_temp1)-1):
+        index=np.argmax(sample[int(rpeak_temp1[i]-np.ceil(fs/10)):int(rpeak_temp1[i]+np.ceil(fs/10)+1)])
+        rpeak_temp1[i]=rpeak_temp1[i]-np.ceil(fs/10)+index
+        count=count+1
+    return np.array(rpeak_temp1).astype(np.int64)
+
+
+def detect_rpeak(ecg,fs:int=64,threshold:float=0.5):
```
Comment (twhnat) [inline memphisdataprocessor/preprocessor/rpeak_detect.py]: Document this function

**Review by twhnat — CHANGES_REQUESTED** [review]

**Comment by aungkonazim** [comment]
Hi Dr. Hnat,

I have just completed the requisites according to your latest commit. Can you please review the changes.

Thank you

**Review by twhnat — APPROVED** [review]
This looks a lot better.  I am going to work on merging this and will let you know what additional changes are needed.

**Pull request merged**